core/src/main/scala/org/apache/spark/storage/memory/MemoryStore.scala

```diff
@@ -387,11 +387,18 @@ private[spark] class MemoryStore(
     // the block's actual memory usage has exceeded the unroll memory by a small amount, so we
     // perform one final call to attempt to allocate additional memory if necessary.
     if (keepUnrolling) {
-      serializationStream.close()
-      reserveAdditionalMemoryIfNecessary()
+      serializationStream.flush()
+      if (bbos.size > unrollMemoryUsedByThisBlock) {
+        val amountToRequest = bbos.size - unrollMemoryUsedByThisBlock
```

Comment:
Here, we only need request the `bbos.size - unrollMemoryUsedByThisBlock`. I'm sorry, this mistake maybe introduced by my previous patch [SPARK-21923](https://github.com/apache/spark/pull/19135).